### PR TITLE
refactor: piece_dispatcher considering score of parent peer

### PR DESCRIPTION
**Pull request by bigerous**
Signed-off-by: bigerous <cuidajun.cdj@alibaba-inc.com>

Refactor dfdaemon's piece dispatcher to improve download speed.

## Test Environment 
20 clients download a same 10GB file. Seedpeer don't have cache.
- seedPeer * 1: 4C4G
- scheduler *1 : 4C4G
- dfget(normal peer) * 20 : 1C1G
- scheduler configuration:
  - Filter Parent Count:6 
  - Load Limit: 50 
  - Download Parallel Count: 4
- seedPeer configuration:
  - Load Limit: 50 
- bandwith between peers: 200MB/s

###  1.before commit of [feat: random pieces download](https://github.com/dragonflyoss/Dragonfly2/pull/1918)：
Mainly uploaded by seedPeer，upload traffic peek is 481MB/s
image
image

###  2.after commit of [feat: random pieces download](https://github.com/dragonflyoss/Dragonfly2/pull/1918)：

But it not work as expectedly
image

### 3.refactor peice random dispatcher
 So I refactor piece random dispatcher. The main difference: first randomly choose a peer, then choose it's piece.

More normal peers are uploading. Upload traffic peek is 1.09GB/s, which is a great improve.
image

But seedPeer is still the most heavy-load peer.
image

### 4.Add 10% randomness and 90% score of peer：
Upload traffic peak is closed to totally random dispather
image

But the load of every peer is more balanced
image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: client/daemon/peer/piece_dispatcher.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+		score:      map[string]int64{},
+		downloaded: map[int32]struct{}{},
+		sum:        atomic.NewInt64(0),
+		closed:     false,
+		cond:       sync.NewCond(lock),
+		lock:       lock,
+		log:        log.With("component", "pieceDispatcher"),
+		rand:       rand.New(rand.NewSource(time.Now().Unix())),
+	}
+	return pd
```
Comment (gaius-qi) [inline client/daemon/peer/piece_dispatcher.go:74]: Return `PieceDispatcher` interface.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: client/daemon/peer/piece_dispatcher.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+	Put(req *DownloadPieceRequest)
+	Get() (req *DownloadPieceRequest, err error)
+	Report(result *DownloadPieceResult)
+	Close()
+}
+
+var ErrNoValidPieceTemporarily = errors.New("no valid piece temporarily")
+
+type pieceDispatcher struct {
+	reqMap     map[string][]*DownloadPieceRequest
```
Comment (gaius-qi) [inline client/daemon/peer/piece_dispatcher.go:43]: Naming does not require a map type.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: client/daemon/peer/piece_dispatcher.go (hunk 2 above)
Comment (bigerous) [reply client/daemon/peer/piece_dispatcher.go]: ReqMap is not a naming map. It's key is peerID, value is a slice of piece requests.

**Review by jim3ma — APPROVED** [review]
LGTM

**Review by gaius-qi — APPROVED** [review]
LGTM

**Pull request merged**